### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by brancz**
This adds a number of tests as well as CI configuration and commits the vendor directory (in a separate commit).

@bwplotka @metalmatze @squat @ant31 @kakkoyun

### CODE REVIEW

**Comment by brancz** [comment]
Sorry for the many pushes, I had an inconsistent linter version locally. All good now, and this is ready for a review.

**Comment by bwplotka** [comment]
`0 / 1942 files viewed` :scream: When someone uses vendoring, part of me dies.. Github is too lagy to comment anything :cry: so I will just comment here...

I would vote for not vendoring and pushing vendor in seperate PR (: 

LGTM, just few nits:

* `testutil.Ok` might save some complexity 
* I think we are missing not to replicate compacted case

File: scheme_test.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+			assert: func(ctx context.Context, t *testing.T, originBucket, targetBucket *inmem.Bucket) {
+				for k := range originBucket.Objects() {
+					if !bytes.Equal(originBucket.Objects()[k], targetBucket.Objects()[k]) {
+						t.Fatalf("Object %s not equal in origin and target bucket.", k)
+					}
+				}
+			},
+		},
+		{
+			name: "OnlyUploadsRaw",
```
Comment (bwplotka) [inline scheme_test.go]: What about cases like:
* when you have compacted one? 
* more than one to upload
* label selection?

**Comment by brancz** [comment]
Added further test scenarios. Please take a look :slightly_smiling_face: .

File: scheme_test.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+				testutil.Ok(t, err)
+				_ = originBucket.Upload(ctx, path.Join(ulid.String(), "meta.json"), bytes.NewReader(b))
+				_ = originBucket.Upload(ctx, path.Join(ulid.String(), "chunks", "000001"), bytes.NewReader(nil))
+				_ = originBucket.Upload(ctx, path.Join(ulid.String(), "index"), bytes.NewReader(nil))
+			},
+			assert: func(ctx context.Context, t *testing.T, originBucket, targetBucket *inmem.Bucket) {
+				expected := 0
+				got := len(targetBucket.Objects())
+				if got != expected {
+					t.Fatalf("TargetBucket should have one block made up of three objects replicated. Got %d but expected %d objects.", got, expected)
```
[the remark this answers is not in the document]
Comment (brancz) [reply scheme_test.go]: yeah copy paste

[the remark this answers is not in the document]
Comment (brancz) [reply scheme_test.go]: fixed

[the remark this answers is not in the document]
Comment (brancz) [reply scheme_test.go]: testutil would work as well, just thought that more descriptive test output would be helpful here

**Pull request merged**